openpype/lib/anatomy.py

```diff
@@ -1040,6 +1051,29 @@ def _format(self, orig_template, data):
             )
             final_data[replace_key_dst] = replace_key_src_curly
 
+        for key in function_to_execute:
+            attrs_function = self.key_and_function_pattern.match(key).groups()
+            if len(attrs_function) <= 1:
+                final_data[key] = "{" + key + "}"
+                continue
+
+            replace_key_dst = "-->".join(attrs_function)
+            replace_key_dst_curly = "{" + replace_key_dst + "}"
+            replace_key_src_curly = "{" + key + "}"
+            template = template.replace(
+                replace_key_src_curly, replace_key_dst_curly
+            )
+            function_name = attrs_function[1]
+            function_attrs = attrs_function[2].split(",")
+            try:
+                value = data[attrs_function[0]]
+                new_value = getattr(value, function_name)(*function_attrs)
+                final_data[replace_key_dst] = new_value
+            except AttributeError:
+                missing_required.append(key)
+            except TypeError as e:
+                invalid_required.append(e)
```

Comment:
`invalid_required` should contain dictionary of `{key: type}`. Where `key` is unsolved key and `type` is invalid type of value that was passed through data.

Reply:
TypeError is for the case where the arguments of the method are not correct. Do I raise an error or do another list invalid_argument?

Reply:
The issue is that you're appending exception to list where is expected dictionary.
```
invalid_required.append({key: type(value)})
```

Reply:
I don't see how to integrate this error into invalid_required, for example if the `key` is `asset.upper(BP)`, it will produce the error `TypeError: file <maya console> line 1: upper() takes no arguments (1 given)`